[ClangImporter] Augment type import to support SwiftAttr in type contexts

xedin opened the pull request:
clang-side change: https://github.com/apple/llvm-project/pull/7918

Teach `importType` that `clang::AttributedType` now carries an optional `clang::Attr *`
that could be used to express concurrency attributes in type context. 

This PR adds support for the following:

- `__attribute__((swft_attr(...)))` is recognized in nested type contexts i.e. blocks, pointers etc.
- Lightweight generics now support `Sendable` conformance requirement.

There are a couple of refactoring here that make things easier to handle:

- ImportTypeKind::CF{Un}retainedOutParameter became attributes instead.
- New `CompletionHandlerParameter` import kind has been introduced to
  denote completion handler function parameters.
- `SendableCompletionHandlers` is now handled based on type attributes.

Resolves: rdar://119341589
Resolves: rdar://104246144
Fixes #58150

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[Serialization] Implementation serialization of clang::SwiftAttrAttr

This is an attribute that is attached to `AttributedType` to support
concurrency annotations in type context.

xedin committed:
[ClangImporter] NFC: Convert CF{Un}retainedOutParaeter into `ImportTypeAttr`

xedin committed:
[ClangImporter] Introduce a new kind of parameter - completion handler

xedin committed:
[ClangImporter] Augment `importType` to infer concurrency attributes from types

Support for `swift_attr` has been extended to type contexts.
Such attributes form `clang::AtributedType` that carries an
optional `clang::Attr *`. Teach `importType` has to examine
immediate sugar to determine whether it carries any concurrency
information and, if so, add it to `ImportTypeAttrs` to be
applied after import.

xedin committed:
[ClangImporter] Move `SendableCompletionHandlers` feature handling into type importer

Parameters no longer expose their attributes at their declaration,
all of the attributes are handled uniformly as type attributes.

xedin committed:
[ClangImporter] NFC: Transform `applyTypeAttributes` into `getConcurrencyAttrs`

It has to be reachable to declaration converter to handle attributes
associated with generic parameters.

xedin committed:
[ClangImporter] Account for Sendable attributes on generic parameter requirements

xedin commented:
https://github.com/apple/swift/pull/70693
@swift-ci please test

xedin commented:
https://github.com/apple/llvm-project/pull/7918
@swift-ci please test

xedin commented:
https://github.com/apple/llvm-project/pull/7918
@swift-ci please test source compatibility

rjmccall reviewed (approved):
Okay, this approach seems reasonable to me.

xedin commented:
https://github.com/apple/llvm-project/pull/7918
@swift-ci please smoke test

xedin commented:
https://github.com/apple/llvm-project/pull/7918
@swift-ci please smoke test

xedin commented:
https://github.com/apple/llvm-project/pull/7918
@swift-ci please smoke test

xedin commented:
https://github.com/apple/llvm-project/pull/7918
@swift-ci please smoke test

xedin committed:
[ClangImporter] Augment `has*Attr` methods to find attributes in type positions

`SwiftAttr` can now appear in type positions and `has*Attr` methods
have to account for that when the given declaration is `ParmVarDecl`.

xedin commented:
https://github.com/apple/llvm-project/pull/7918
@swift-ci please smoke test

xedin commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -7196,36 +7196,40 @@ static bool isForeignReferenceType(const clang::QualType type) {
-             return swiftAttr->getAttribute() == "import_owned";
-           return false;
-         });
+  return hasSwiftAttribute(decl, "import_owned");
```
cc @egorzhdan
